File: java-samples/referencestates-sanctionsbody/contracts/src/main/java/com/r3/developers/samples/referencestate/contracts/SanctionListContract.java
package com.r3.developers.samples.referencestate.contracts;

import com.r3.developers.samples.referencestate.states.SanctionList;
import net.corda.v5.base.exceptions.CordaRuntimeException;
import net.corda.v5.ledger.utxo.Command;
import net.corda.v5.ledger.utxo.Contract;
import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction;
import org.jetbrains.annotations.NotNull;

public class SanctionListContract implements Contract {

    public static class Create implements Command { }

    public static class Update implements Command { }

    static final String REQUIRES_ONE_COMMAND = "The transaction requires exactly one command";
    static final String CREATE_REQUIRES_ZERO_INPUTS = "When creating a sanctions list there should be no inputs";
    static final String CREATE_REQUIRES_ONE_OUTPUT = "When creating a sanctions list there should be exactly one output";
    static final String CREATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT =
            "When creating a sanctions list there should be one output of type SanctionedEntities";
    static final String CREATE_REQUIRES_ISSUER_SIGNATURE = "The issuer of the sanctions list must sign";

    static final String UPDATE_REQUIRES_ONE_INPUT = "There must be exactly one input when updating Sanctions List";
    static final String UPDATE_REQUIRES_ONE_OUTPUT = "There must be exactly one output when updating Sanctions List";
    static final String UPDATE_REQUIRES_ONE_SANCTION_ENTITY_INPUT =
            "There must be exactly one input Sanctions List when updating";
    static final String UPDATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT =
            "There must be exactly one input Sanctions List when updating";
    static final String UPDATE_ISSUER_SHOULD_NOT_CHANGE =
            "The issuer must remain the same across an update";


    @Override
    public void verify(@NotNull UtxoLedgerTransaction transaction) {

        requireThat(!transaction.getCommands().isEmpty(), REQUIRES_ONE_COMMAND);
        Command command = transaction.getCommands().get(0);

        if (command instanceof Create) {
            requireThat(transaction.getInputContractStates().isEmpty(), CREATE_REQUIRES_ZERO_INPUTS);
            requireThat(transaction.getOutputContractStates().size() ==1, CREATE_REQUIRES_ONE_OUTPUT);
            requireThat(transaction.getOutputStates(SanctionList.class).size() == 1,
                    CREATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT);

            SanctionList out = transaction.getOutputStates(SanctionList.class).get(0);
            requireThat(transaction.getSignatories().contains(out.getIssuer().getLedgerKey()), CREATE_REQUIRES_ISSUER_SIGNATURE);

        } else if (command instanceof Update) {
            requireThat(transaction.getInputContractStates().size() == 1, UPDATE_REQUIRES_ONE_INPUT);
            requireThat(transaction.getOutputContractStates().size() == 1, UPDATE_REQUIRES_ONE_OUTPUT);

            requireThat(transaction.getInputStates(SanctionList.class).size() == 1,
                    UPDATE_REQUIRES_ONE_SANCTION_ENTITY_INPUT);
            requireThat(transaction.getOutputStates(SanctionList.class).size() == 1,
                    UPDATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT);

            SanctionList input = transaction.getInputStates(SanctionList.class).get(0);
            SanctionList output = transaction.getOutputStates(SanctionList.class).get(0);
            requireThat(input.getIssuer().getName().equals(output.getIssuer().getName()),
                    UPDATE_ISSUER_SHOULD_NOT_CHANGE);
        }
    }

    private void requireThat(boolean asserted, String errorMessage) {
        if(!asserted) {
            throw new CordaRuntimeException("Failed requirement: " + errorMessage);
        }
    }
}

File: kotlin-samples/referencestates-sanctionsbody/contracts/src/main/kotlin/com/r3/developers/samples/referencestate/contracts/SanctionListContract.kt
package com.r3.developers.samples.referencestate.contracts

import com.r3.developers.samples.referencestate.states.SanctionList
import net.corda.v5.base.exceptions.CordaRuntimeException
import net.corda.v5.ledger.utxo.Command
import net.corda.v5.ledger.utxo.Contract
import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction

class SanctionListContract :Contract {

    internal companion object {
        const val REQUIRES_ONE_COMMAND = "The transaction requires exactly one command"
        const val CREATE_REQUIRES_ZERO_INPUTS = "When creating a sanctions list there should be no inputs"
        const val CREATE_REQUIRES_ONE_OUTPUT = "When creating a sanctions list there should be exactly one output"
        const val CREATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT = "When creating a sanctions list there should be one output of type SanctionedEntities"
        const val CREATE_REQUIRES_ISSUER_SIGNATURE = "The issuer of the sanctions list must sign"

        const val UPDATE_REQUIRES_ONE_INPUT = "There must be exactly one input when updating Sanctions List"
        const val UPDATE_REQUIRES_ONE_OUTPUT = "There must be exactly one output when updating Sanctions List"
        const val UPDATE_REQUIRES_ONE_SANCTION_ENTITY_INPUT = "There must be exactly one input Sanctions List when updating"
        const val UPDATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT = "There must be exactly one input Sanctions List when updating"
        const val UPDATE_ISSUER_SHOULD_NOT_CHANGE = "The issuer must remain the same across an update"
    }

    interface SanctionListCommand : Command {
        class Create: SanctionListCommand
        class Update: SanctionListCommand
    }

    override fun verify(transaction: UtxoLedgerTransaction) {

        val command = transaction.getCommands(SanctionListCommand::class.java).singleOrNull()  ?: throw CordaRuntimeException(REQUIRES_ONE_COMMAND)
        when(command){
            is SanctionListCommand.Create -> {
                CREATE_REQUIRES_ZERO_INPUTS using (transaction.inputContractStates.isEmpty())
                CREATE_REQUIRES_ONE_OUTPUT using (transaction.outputContractStates.size == 1)
                CREATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT using (transaction.getOutputStates(SanctionList::class.java).size == 1)
                val output = transaction.getOutputStates(SanctionList::class.java)[0]
                CREATE_REQUIRES_ISSUER_SIGNATURE using (transaction.signatories.contains(output.issuer.ledgerKey))
            }
            is SanctionListCommand.Update -> {
                UPDATE_REQUIRES_ONE_INPUT using (transaction.inputContractStates.size == 1)
                UPDATE_REQUIRES_ONE_OUTPUT using (transaction.outputContractStates.size == 1)
                UPDATE_REQUIRES_ONE_SANCTION_ENTITY_INPUT using (transaction.getInputStates(SanctionList::class.java).size == 1)
                UPDATE_REQUIRES_ONE_SANCTION_ENTITY_OUTPUT using (transaction.getOutputStates(SanctionList::class.java).size == 1)
                val input = transaction.getInputStates(SanctionList::class.java)[0]
                val output = transaction.getOutputStates(SanctionList::class.java)[0]
                UPDATE_ISSUER_SHOULD_NOT_CHANGE using (input.issuer.name==output.issuer.name)
            }else -> {
                throw CordaRuntimeException("Invalid Command")
            }
        }
    }
    // Helper function to allow writing constraints in the Corda 4 '"text" using (boolean)' style
    private infix fun String.using(expr: Boolean) {
        if (!expr) throw CordaRuntimeException("Failed requirement: $this")
    }

    // Helper function to allow writing constraints in '"text" using {lambda}' style where the last expression
    // in the lambda is a boolean.
    private infix fun String.using(expr: () -> Boolean) {
        if (!expr.invoke()) throw CordaRuntimeException("Failed requirement: $this")
    }
}